spec/abstract/SchedContext_A.thy

```diff
@@ -101,6 +101,32 @@ definition
 where
   "set_refills sc_ptr refills = set_sc_obj_ref sc_refills_update sc_ptr refills"
 
+definition
+  update_refill_hd :: "obj_ref \<Rightarrow> (refill \<Rightarrow> refill) \<Rightarrow> (unit, 'z::state_ext) s_monad"
+where
+  "update_refill_hd sc_ptr f = do
+     refills \<leftarrow> get_refills sc_ptr;
+     set_refills sc_ptr (f (hd refills) # (tl refills))
+   od"
+
+definition
+  set_refill_hd :: "obj_ref \<Rightarrow> refill \<Rightarrow> (unit, 'z::state_ext) s_monad"
+where
+  "set_refill_hd sc_ptr new = update_refill_hd sc_ptr (\<lambda>_. new)"
+
+definition
+  update_refill_tl :: "obj_ref \<Rightarrow> (refill \<Rightarrow> refill) \<Rightarrow> (unit, 'z::state_ext) s_monad"
+where
```

Comment:
These functions were just ripped from the Haskell, and they are fairly close to what the C has. 

I'm really only using these functions at the moment to improve the readability of the abstract spec (I've got a bundle `refill_budget_check_defs` which just unfolds these, and I think it's unfolded everywhere else, too) though I suppose if it's confusing, it's not actually helping readability.

I got quite used to these functions when working on the `corres` rules for the refill functions. I quite like them. We also have the abbreviation `refill_tl` in AInvs for the last element of the refill list. Seeing that they would really only appear if someone is looking into the refill logic, I think it would be best to leave them here